shell/components/mic-input.js

```diff
@@ -0,0 +1,141 @@
+      document.addEventListener('keyup', this._onKeyup.bind(this));
+    }
+  }
+  _update(props, state) {
+    if (state.active && !state.recognizing) {
+      this._start();
+    }
+    if (!state.active && state.recognizing) {
+      this._stop();
+    }
+  }
+  _render(props, state) {
+    return state;
+  }
+  _supportsSpeech() {
+    return ('webkitSpeechRecognition' in window);
+  }
+  _initSpeech(state) {
+    const recognition = new window.webkitSpeechRecognition();
+    recognition.continuous = true;
+    recognition.interimResults = true;
+    recognition.onstart = this._onRecognitionStart.bind(this);
+    recognition.onerror = this._onRecognitionError.bind(this);
+    recognition.onend = this._onRecognitionEnd.bind(this);
+    recognition.onresult = this._onRecognitionResult.bind(this);
+    this._setState({recognition});
+  }
+  start() {
+    const {recognition, recognizing, finalTranscript} = this._state;
+    if (recognition) {
+      this._start();
+    } // else complain
+  }
+  _start() {
+    const {recognition, recognizing, duration, finalTranscript} = this._state;
+    if (recognizing) {
+      recognition.stop();
+    } else {
+      recognition.start();
+      this._setState({
+        finalTranscript: '',
+        ignoreOnEnd: false
+      });
+    }
+  }
+  _stop() {
+    const {recognition, recognizing, duration, finalTranscript} = this._state;
+    if (recognizing) {
+      recognition.stop();
+    }
+  }
+  _onStartClick() {
+    this.start();
+  }
+  _onRecognitionStart() {
+    this._setState({recognizing: true});
+    this._fire('start');
+  }
+  _onRecognitionError() {
+    if (event.error == 'no-speech') {
+      // TODO: something?
+    }
+    if (event.error == 'audio-capture') {
+      // TODO: something?
+    }
+    if (event.error == 'not-allowed') {
+      // TODO: something?
+    }
+    this._setState({ignoreOnEnd: true});
+  }
+  _onRecognitionResult() {
+    let {ignoreOnEnd, finalTranscript} = this._state;
+    let interimTranscript = '';
+    for (let i = event.resultIndex; i < event.results.length; ++i) {
+      if (event.results[i].isFinal) {
+        finalTranscript += event.results[i][0].transcript;
+      } else {
+        interimTranscript += event.results[i][0].transcript;
+      }
+    }
+    this._setState({interimTranscript, finalTranscript});
+    const transcript = `${finalTranscript}${interimTranscript}`;
+    this.value = transcript;
+    this._fire('result', transcript);
+  }
+  _onRecognitionEnd() {
+    const {ignoreOnEnd, finalTranscript} = this._state;
+    this._setState({recognizing: false});
+    // TODO(sjmiles): these conditionals in original demo code, leaving here in case I figure out why
+    /*
+    if (!ignoreOnEnd) {
+      if (!finalTranscript) {
+      } else {
+      }
+    }
+    */
+    this.value = finalTranscript;
+    this._fire('end', finalTranscript);
+  }
+  _onKeydown(e) {
+    const name = e.target.localName;
```

Comment:
nit: this and next handler are same, could factor to single method they both call with varying boolean.